File: backend/app/schemas/voice_registry.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class VoiceProviderResponse(_StrictModel):
    key: str
    name: str
    status: Literal["active", "planned"]
    supports_managed_credentials: bool
    supports_byok: bool


class ParameterSpecResponse(_StrictModel):
    supported: bool
    type: Literal["number", "integer", "boolean", "string", "enum"] | None = None
    min: float | None = None
    max: float | None = None
    step: float | None = None
    default: Any = None
    options: tuple[str, ...] | None = None
    advanced: bool = False


class VoiceModelResponse(_StrictModel):
    id: str
    provider_key: str
    key: str
    name: str
    execution_model_id: str
    model_type: Literal["stt", "llm", "tts", "realtime"]
    implementation_status: Literal["planned", "available", "deprecated"]
    capabilities: dict[str, bool]
    parameters: dict[str, ParameterSpecResponse]
    external_voice_providers: tuple[str, ...] = ()

File: frontend/types/voice-registry.ts
export type VoiceProviderStatus = 'active' | 'planned';
export type VoiceModelType = 'stt' | 'llm' | 'tts' | 'realtime';
export type VoiceModelImplementationStatus = 'planned' | 'available' | 'deprecated';

export type VoiceProviderResponse = {
  key: string;
  name: string;
  status: VoiceProviderStatus;
  supports_managed_credentials: boolean;
  supports_byok: boolean;
};

export type ParameterSpecType = 'number' | 'integer' | 'boolean' | 'string' | 'enum';

export type ParameterSpecResponse = {
  supported: boolean;
  type?: ParameterSpecType | null;
  min?: number | null;
  max?: number | null;
  step?: number | null;
  default?: unknown;
  options?: string[] | null;
  advanced?: boolean;
};

export type VoiceModelResponse = {
  id: string;
  provider_key: string;
  key: string;
  name: string;
  execution_model_id: string;
  model_type: VoiceModelType;
  implementation_status: VoiceModelImplementationStatus;
  capabilities: Record<string, boolean>;
  parameters: Record<string, ParameterSpecResponse>;
  external_voice_providers?: string[];
};
